### DISCUSSION TITLE
Uvicorn with Thread handle KeyboardInterrupt

### DISCUSSION THREAD

**Discussion by cxheng315** [fastapi/fastapi#6210] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, APIRouter, Query
import uvicorn
import threading

def run_background_task():
        
    while True:
        try:
            print("Auto Every 5 minute Update Data")
            fetch_data()
            time.sleep(60*60)
        except Exception as e:
            print(e)

if __name__ == "__main__":
    try:
        stop_event = asyncio.Event()
        thread = threading.Thread(target=run_background_task)
        thread.start()
        uvicorn.run(app, host="0.0.0.0", port=8000, log_level="info")
    except KeyboardInterrupt:
        # Set the stop event to stop the server and the background task
        # Wait for the background task to stop
        thread.join()
    except Exception as e:
        print(e)
```

### Description

After I pressed Ctrl + C, the KeyboardInterrupt, the uvicorn.run(), server stopped, but the thread unable to terminate.
Seems like uvicorn.run() will handle KeyboardInterrupt internally, so I'm unable to terminate the thread.
How should I do to terminate the uvicorn server and thread the same time when I entered Ctrl + C

### Operating System

Windows

### Operating System Details

Windows 11

### FastAPI Version

0.88.0

### Python Version

Python 3.10.6

**Comment by NewSouthMjos** [comment, marked as answer]
Fast solution: use fastapi-utils [repeat_every](https://fastapi-utils.davidmontague.xyz/user-guide/repeated-tasks/)

**Comment by jgould22** [comment]
Uvicorn adds a signal handler to the event loop.

 https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.add_signal_handler

Also I will have to assume this is a toy example because the example background task will never quit as written. 

https://docs.python.org/3/library/signal.html#signals-and-threads

Perhaps you can pass your thread an [event](https://docs.python.org/3/library/threading.html#event-objects) and register a callback that sets your event. Then you can check `is_set` in your thread.

**Comment by cxheng315** [comment]
> Fast solution: use fastapi-utils [repeat_every](https://fastapi-utils.davidmontague.xyz/user-guide/repeated-tasks/)

Problem solved, thank you so much.
